File: apps/gsd-cloud-monitor/Sources/GSDCloudMonitorCore/RuntimeTelemetry.swift
import Foundation

public enum RuntimeConnectionState: String, Codable, Sendable {
  case connecting
  case connected
  case reconnecting
  case error
  case stopped
  case stale
}

public func telemetryUnavailableState(
  validatedProcessIsRunning: Bool?
) -> RuntimeConnectionState {
  validatedProcessIsRunning == false ? .stopped : .stale
}

public struct TelemetryUnavailableStateTracker: Sendable {
  private var validatedProcessIsRunning: Bool?

  public var connectionState: RuntimeConnectionState {
    telemetryUnavailableState(validatedProcessIsRunning: validatedProcessIsRunning)
  }

  public init() {}

  public mutating func recordProcessValidation(isRunning: Bool?) {
    validatedProcessIsRunning = isRunning
  }

  public mutating func reset() {
    validatedProcessIsRunning = nil
  }
}

public struct TelemetryFreshnessTracker: Sendable {
  private var lastUpdatedAt: Date?
  private var missedObservations = 0

  public init() {}

  public mutating func connectionState(
    reportedState: RuntimeConnectionState,
    updatedAt: Date,
    processIsRunning: Bool,
    now _: Date = Date()
  ) -> RuntimeConnectionState {
    guard processIsRunning else {
      reset()
      return .stopped
    }
    if lastUpdatedAt == updatedAt {
      missedObservations += 1
    } else {
      lastUpdatedAt = updatedAt
      missedObservations = 0
    }
    return missedObservations >= 2 ? .stale : reportedState
  }

  public mutating func reset() {
    lastUpdatedAt = nil
    missedObservations = 0
  }
}

public enum RuntimeProjectState: String, Codable, Sendable {
  case idle
  case active
  case error
}

public enum RuntimeActivityOutcome: String, Codable, Sendable {
  case success
  case error
  case cancelled
}

public struct RuntimeProjectTelemetry: Decodable, Identifiable, Sendable {
  public var id: String { path }

  public let alias: String
  public let path: String
  public let repoIdentity: String
  public let remoteLabel: String?
  public let state: RuntimeProjectState
  public let activeRequests: Int
  public let activeTools: [String]
  public let requestCount: Int
  public let errorCount: Int
  public let receivedBytes: Int64
  public let sentBytes: Int64
  public let lastTool: String?
  public let lastActivityAt: Date?

  private enum CodingKeys: String, CodingKey {
    case alias
    case path
    case repoIdentity = "repo_identity"
    case remoteLabel = "remote_label"
    case state
    case activeRequests = "active_requests"
    case activeTools = "active_tools"
    case requestCount = "request_count"
    case errorCount = "error_count"
    case receivedBytes = "received_bytes"
    case sentBytes = "sent_bytes"
    case lastTool = "last_tool"
    case lastActivityAt = "last_activity_at"
  }

  public init(from decoder: Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)
    alias = try values.decode(String.self, forKey: .alias)
    path = try values.decode(String.self, forKey: .path)
    repoIdentity = try values.decode(String.self, forKey: .repoIdentity)
    remoteLabel = try values.decodeIfPresent(String.self, forKey: .remoteLabel)
    state = try values.decode(RuntimeProjectState.self, forKey: .state)
    activeRequests = try values.decode(Int.self, forKey: .activeRequests)
    activeTools = try values.decodeIfPresent([String].self, forKey: .activeTools) ?? []
    requestCount = try values.decode(Int.self, forKey: .requestCount)
    errorCount = try values.decode(Int.self, forKey: .errorCount)
    receivedBytes = try values.decode(Int64.self, forKey: .receivedBytes)
    sentBytes = try values.decode(Int64.self, forKey: .sentBytes)
    lastTool = try values.decodeIfPresent(String.self, forKey: .lastTool)
    lastActivityAt = try values.decodeIfPresent(Date.self, forKey: .lastActivityAt)
  }
}

public struct RuntimeActivity: Decodable, Identifiable, Sendable {
  public var id: String { requestID }

  public let requestID: String
  public let projectAlias: String?
  public let projectPath: String?
  public let toolName: String
  public let outcome: RuntimeActivityOutcome
  public let durationMs: Int
  public let at: Date
  public let error: String?

  private enum CodingKeys: String, CodingKey {
    case requestID = "request_id"
    case projectAlias = "project_alias"
    case projectPath = "project_path"
    case toolName = "tool_name"
    case outcome
    case durationMs = "duration_ms"
    case at
    case error
  }

  public func belongs(to project: RuntimeProjectTelemetry) -> Bool {
    if let projectPath { return projectPath == project.path }
    return projectAlias == project.alias
  }
}

public struct RuntimeTelemetry: Decodable, Sendable {
  public let version: Int
  public let pid: Int32
  public let state: RuntimeConnectionState
  public let gatewayURL: URL
  public let runtimeID: String?
  public let runtimeName: String?
  public let startedAt: Date
  public let connectedAt: Date?
  public let updatedAt: Date
  public let lastError: String?
  public let connectionAttempts: Int
  public let reconnects: Int
  public let receivedMessages: Int
  public let sentMessages: Int
  public let receivedBytes: Int64
  public let sentBytes: Int64
  public let activeRequests: Int
  public let projects: [RuntimeProjectTelemetry]
  public let recentActivity: [RuntimeActivity]

  private enum CodingKeys: String, CodingKey {
    case version
    case pid
    case state
    case gatewayURL = "gateway_url"
    case runtimeID = "runtime_id"
    case runtimeName = "runtime_name"
    case startedAt = "started_at"
    case connectedAt = "connected_at"
    case updatedAt = "updated_at"
    case lastError = "last_error"
    case connectionAttempts = "connection_attempts"
    case reconnects
    case receivedMessages = "received_messages"
    case sentMessages = "sent_messages"
    case receivedBytes = "received_bytes"
    case sentBytes = "sent_bytes"
    case activeRequests = "active_requests"
    case projects
    case recentActivity = "recent_activity"
  }

  public init(from decoder: Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)
    version = try values.decode(Int.self, forKey: .version)
    pid = try values.decode(Int32.self, forKey: .pid)
    state = try values.decode(RuntimeConnectionState.self, forKey: .state)
    gatewayURL = try values.decode(URL.self, forKey: .gatewayURL)
    runtimeID = try values.decodeIfPresent(String.self, forKey: .runtimeID)
    runtimeName = try values.decodeIfPresent(String.self, forKey: .runtimeName)
    startedAt = try values.decode(Date.self, forKey: .startedAt)
    connectedAt = try values.decodeIfPresent(Date.self, forKey: .connectedAt)
    updatedAt = try values.decode(Date.self, forKey: .updatedAt)
    lastError = try values.decodeIfPresent(String.self, forKey: .lastError)
    connectionAttempts = try values.decode(Int.self, forKey: .connectionAttempts)
    reconnects = try values.decode(Int.self, forKey: .reconnects)
    receivedMessages = try values.decode(Int.self, forKey: .receivedMessages)
    sentMessages = try values.decode(Int.self, forKey: .sentMessages)
    receivedBytes = try values.decode(Int64.self, forKey: .receivedBytes)
    sentBytes = try values.decode(Int64.self, forKey: .sentBytes)
    activeRequests = try values.decode(Int.self, forKey: .activeRequests)
    projects = try values.decodeIfPresent([RuntimeProjectTelemetry].self, forKey: .projects) ?? []
    recentActivity =
      try values.decodeIfPresent([RuntimeActivity].self, forKey: .recentActivity) ?? []
  }
}

public struct TrafficCounters: Equatable, Sendable {
  public let receivedBytes: Int64
  public let sentBytes: Int64

  public init(receivedBytes: Int64, sentBytes: Int64) {
    self.receivedBytes = receivedBytes
    self.sentBytes = sentBytes
  }

  public func rate(since previous: TrafficCounters, elapsed: TimeInterval) -> TrafficRate {
    guard elapsed > 0 else { return .zero }
    let receivedDelta = max(0, receivedBytes - previous.receivedBytes)
    let sentDelta = max(0, sentBytes - previous.sentBytes)
    return TrafficRate(
      receivedBytesPerSecond: Double(receivedDelta) / elapsed,
      sentBytesPerSecond: Double(sentDelta) / elapsed
    )
  }
}

public struct TrafficRate: Equatable, Sendable {
  public let receivedBytesPerSecond: Double
  public let sentBytesPerSecond: Double

  public init(receivedBytesPerSecond: Double, sentBytesPerSecond: Double) {
    self.receivedBytesPerSecond = receivedBytesPerSecond
    self.sentBytesPerSecond = sentBytesPerSecond
  }

  public static let zero = TrafficRate(receivedBytesPerSecond: 0, sentBytesPerSecond: 0)
}

public struct RuntimeTelemetryReader: Sendable {
  public init() {}

  public func load(from fileURL: URL) throws -> RuntimeTelemetry {
    let data = try Data(contentsOf: fileURL)
    let decoder = JSONDecoder()
    decoder.dateDecodingStrategy = .custom { decoder in
      let container = try decoder.singleValueContainer()
      let value = try container.decode(String.self)
      if let date = Self.fractionalDateFormatter.date(from: value)
        ?? Self.dateFormatter.date(from: value)
      {
        return date
      }
      throw DecodingError.dataCorruptedError(
        in: container,
        debugDescription: "Invalid ISO-8601 date: \(value)"
      )
    }
    let telemetry = try decoder.decode(RuntimeTelemetry.self, from: data)
    guard telemetry.version == 1 else {
      throw RuntimeTelemetryError.unsupportedVersion(telemetry.version)
    }
    return telemetry
  }

  private static let fractionalDateFormatter: ISO8601DateFormatter = {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return formatter
  }()

  private static let dateFormatter = ISO8601DateFormatter()
}

public enum RuntimeTelemetryError: Error, Equatable {
  case unsupportedVersion(Int)
}

extension RuntimeTelemetry {
  public var trafficCounters: TrafficCounters {
    TrafficCounters(receivedBytes: receivedBytes, sentBytes: sentBytes)
  }
}

extension RuntimeProjectTelemetry {
  public var trafficCounters: TrafficCounters {
    TrafficCounters(receivedBytes: receivedBytes, sentBytes: sentBytes)
  }

  public var activeToolSummary: String {
    activeTools.isEmpty ? "\(activeRequests) active" : activeTools.joined(separator: ", ")
  }
}

File: packages/gsd-cloud/src/runtime-telemetry.ts
// Project/App: Open GSD
// File Purpose: Persist token-free cloud runtime status and traffic counters for local monitors.
import { readFileSync, unlinkSync } from "node:fs";
import { chmod, mkdir, rename, rm, writeFile } from "node:fs/promises";
import { dirname } from "node:path";
import type { AdvertisedProject } from "./executors/executor.js";
import { runtimeArtifactPath } from "./runtime-artifacts.js";

export type RuntimeConnectionState = "connecting" | "connected" | "reconnecting" | "error" | "stopped";
export type RuntimeProjectState = "idle" | "active" | "error";
export type RuntimeRequestOutcome = "success" | "error" | "cancelled";

export interface RuntimeProjectTelemetry {
  alias: string;
  path: string;
  repo_identity: string;
  remote_label?: string;
  state: RuntimeProjectState;
  active_requests: number;
  active_tools: string[];
  request_count: number;
  error_count: number;
  received_bytes: number;
  sent_bytes: number;
  last_tool: string | null;
  last_activity_at: string | null;
}

export interface RuntimeActivityTelemetry {
  request_id: string;
  project_alias?: string;
  project_path?: string;
  tool_name: string;
  outcome: RuntimeRequestOutcome;
  duration_ms: number;
  at: string;
  error?: string;
}

export interface RuntimeTelemetryStatus {
  version: 1;
  pid: number;
  state: RuntimeConnectionState;
  gateway_url: string;
  runtime_id?: string;
  runtime_name?: string;
  started_at: string;
  connected_at: string | null;
  updated_at: string;
  last_error: string | null;
  connection_attempts: number;
  reconnects: number;
  received_messages: number;
  sent_messages: number;
  received_bytes: number;
  sent_bytes: number;
  active_requests: number;
  projects: RuntimeProjectTelemetry[];
  recent_activity: RuntimeActivityTelemetry[];
}

export interface RuntimeRequestStarted {
  requestId: string;
  projectAlias?: string;
  projectPath?: string;
  toolName: string;
  receivedBytes: number;
}

export interface RuntimeRequestFinished {
  requestId: string;
  projectAlias?: string;
  projectPath?: string;
  toolName: string;
  durationMs: number;
  outcome: RuntimeRequestOutcome;
  error?: string;
}

export interface RuntimeTelemetryReporter {
  connecting(): void;
  connected(): void;
  disconnected(error?: string): void;
  socketError(error: string): void;
  received(text: string): void;
  sent(text: string, projectPath?: string): void;
  projectsAdvertised(projects: AdvertisedProject[]): void;
  requestStarted(request: RuntimeRequestStarted): void;
  requestFinished(request: RuntimeRequestFinished): void;
  stopped(): void;
  failed?(): void;
  flush?(): Promise<void>;
}

interface RuntimeTelemetryMetadata {
  gatewayUrl: string;
  runtimeId?: string;
  runtimeName?: string;
}

export const noopRuntimeTelemetry: RuntimeTelemetryReporter = {
  connecting(): void {},
  connected(): void {},
  disconnected(): void {},
  socketError(): void {},
  received(): void {},
  sent(): void {},
  projectsAdvertised(): void {},
  requestStarted(): void {},
  requestFinished(): void {},
  stopped(): void {},
  failed(): void {},
};

export class RuntimeTelemetryStore implements RuntimeTelemetryReporter {
  private readonly path: string;
  private readonly activeRequests = new Map<string, { projectPath: string; toolName: string }>();
  private status: RuntimeTelemetryStatus;
  private livenessTimer: ReturnType<typeof setInterval> | undefined;
  private persistTimer: ReturnType<typeof setTimeout> | undefined;
  private persistPromise = Promise.resolve();

  constructor(configPath: string, metadata: RuntimeTelemetryMetadata) {
    const now = new Date().toISOString();
    this.path = runtimeTelemetryPath(configPath);
    this.status = {
      version: 1,
      pid: process.pid,
      state: "connecting",
      gateway_url: credentialFreeGatewayLabel(metadata.gatewayUrl),
      ...(metadata.runtimeId ? { runtime_id: metadata.runtimeId } : {}),
      ...(metadata.runtimeName ? { runtime_name: metadata.runtimeName } : {}),
      started_at: now,
      connected_at: null,
      updated_at: now,
      last_error: null,
      connection_attempts: 0,
      reconnects: 0,
      received_messages: 0,
      sent_messages: 0,
      received_bytes: 0,
      sent_bytes: 0,
      active_requests: 0,
      projects: [],
      recent_activity: [],
    };
  }

  connecting(): void {
    if (!this.livenessTimer) {
      this.livenessTimer = setInterval(() => this.persist(), 1_000);
      this.livenessTimer.unref();
    }
    this.status.state = this.status.connection_attempts === 0 ? "connecting" : "reconnecting";
    this.status.connection_attempts += 1;
    this.persist();
  }

  connected(): void {
    this.status.state = "connected";
    this.status.connected_at = new Date().toISOString();
    this.status.last_error = null;
    this.persist();
  }

  disconnected(error?: string): void {
    if (this.status.connected_at && this.status.state !== "reconnecting") {
      this.status.reconnects += 1;
    }
    this.status.state = "reconnecting";
    this.status.last_error = error ?? this.status.last_error ?? "connection closed";
    this.persist();
  }

  socketError(error: string): void {
    this.status.state = "error";
    this.status.last_error = error;
    this.persist();
  }

  received(text: string): void {
    this.status.received_messages += 1;
    this.status.received_bytes += Buffer.byteLength(text);
    this.persist();
  }

  sent(text: string, projectPath?: string): void {
    this.status.sent_messages += 1;
    const bytes = Buffer.byteLength(text);
    this.status.sent_bytes += bytes;
    const project = projectPath ? this.findProject(projectPath) : undefined;
    if (project) project.sent_bytes += bytes;
    this.persist();
  }

  projectsAdvertised(projects: AdvertisedProject[]): void {
    const existingProjects = new Map(
      this.status.projects.map((project) => [project.path, project]),
    );
    this.status.projects = projects.map((project) => {
      const existing = existingProjects.get(project.path);
      const remoteLabel = project.remoteLabel
        ? credentialFreeRemoteLabel(project.remoteLabel)
        : undefined;
      return {
        alias: project.alias,
        path: project.path,
        repo_identity: project.repoIdentity,
        ...(remoteLabel ? { remote_label: remoteLabel } : {}),
        state: existing?.state ?? "idle",
        active_requests: existing?.active_requests ?? 0,
        active_tools: existing?.active_tools ?? [],
        request_count: existing?.request_count ?? 0,
        error_count: existing?.error_count ?? 0,
        received_bytes: existing?.received_bytes ?? 0,
        sent_bytes: existing?.sent_bytes ?? 0,
        last_tool: existing?.last_tool ?? null,
        last_activity_at: existing?.last_activity_at ?? null,
      };
    });
    this.persist();
  }

  requestStarted(request: RuntimeRequestStarted): void {
    this.status.active_requests += 1;
    const project = this.findProject(request.projectPath, request.projectAlias);
    if (project) {
      this.activeRequests.set(request.requestId, {
        projectPath: project.path,
        toolName: request.toolName,
      });
      project.state = "active";
      project.active_requests += 1;
      project.active_tools = this.activeToolsForProject(project.path);
      project.received_bytes += request.receivedBytes;
      project.last_tool = request.toolName;
      project.last_activity_at = new Date().toISOString();
    }
    this.persist();
  }

  requestFinished(request: RuntimeRequestFinished): void {
    this.status.active_requests = Math.max(0, this.status.active_requests - 1);
    const now = new Date().toISOString();
    const project = this.findProject(request.projectPath, request.projectAlias);
    this.activeRequests.delete(request.requestId);
    if (project) {
      project.active_requests = Math.max(0, project.active_requests - 1);
      project.active_tools = this.activeToolsForProject(project.path);
      project.request_count += 1;
      project.last_tool = request.toolName;
      project.last_activity_at = now;
      if (request.outcome === "error") project.error_count += 1;
      if (request.outcome === "error") {
        project.state = "error";
      } else if (project.active_requests > 0) {
        project.state = "active";
      } else {
        project.state = "idle";
      }
    }
    this.status.recent_activity.push({
      request_id: request.requestId,
      ...(request.projectAlias ? { project_alias: request.projectAlias } : {}),
      ...(request.projectPath ? { project_path: request.projectPath } : {}),
      tool_name: request.toolName,
      outcome: request.outcome,
      duration_ms: request.durationMs,
      at: now,
      ...(request.error ? { error: request.error } : {}),
    });
    if (this.status.recent_activity.length > 50) {
      this.status.recent_activity.splice(0, this.status.recent_activity.length - 50);
    }
    this.persist();
  }

  stopped(): void {
    if (this.livenessTimer) clearInterval(this.livenessTimer);
    this.livenessTimer = undefined;
    this.status.state = "stopped";
    this.status.active_requests = 0;
    this.activeRequests.clear();
    for (const project of this.status.projects) {
      project.active_requests = 0;
      project.active_tools = [];
      project.state = "idle";
    }
    void this.flush();
  }

  failed(): void {
    if (this.livenessTimer) clearInterval(this.livenessTimer);
    this.livenessTimer = undefined;
  }

  async flush(): Promise<void> {
    if (this.persistTimer) {
      clearTimeout(this.persistTimer);
      this.persistTimer = undefined;
    }
    this.enqueuePersist();
    await this.persistPromise;
  }

  private findProject(path?: string, alias?: string): RuntimeProjectTelemetry | undefined {
    if (path) return this.status.projects.find((project) => project.path === path);
    if (alias) return this.status.projects.find((project) => project.alias === alias);
    if (this.status.projects.length === 1) return this.status.projects[0];
    return undefined;
  }

  private activeToolsForProject(projectPath: string): string[] {
    return [...this.activeRequests.values()]
      .filter((request) => request.projectPath === projectPath)
      .map((request) => request.toolName);
  }

  private persist(): void {
    this.status.updated_at = new Date().toISOString();
    if (this.persistTimer) return;
    this.persistTimer = setTimeout(() => {
      this.persistTimer = undefined;
      this.enqueuePersist();
    }, 25);
    this.persistTimer.unref();
  }

  private enqueuePersist(): void {
    this.status.updated_at = new Date().toISOString();
    const snapshot = `${JSON.stringify(this.status, null, 2)}\n`;
    this.persistPromise = this.persistPromise.then(() => this.writeSnapshot(snapshot));
  }

  private async writeSnapshot(snapshot: string): Promise<void> {
    const temporaryPath = `${this.path}.${process.pid}.tmp`;
    try {
      await mkdir(dirname(this.path), { recursive: true });
      await writeFile(temporaryPath, snapshot, { encoding: "utf8", mode: 0o600 });
      await chmod(temporaryPath, 0o600);
      await rename(temporaryPath, this.path);
    } catch {
      await rm(temporaryPath, { force: true }).catch(() => undefined);
    }
  }
}

export function readRuntimeTelemetry(configPath: string): RuntimeTelemetryStatus | null {
  try {
    const value = JSON.parse(readFileSync(runtimeTelemetryPath(configPath), "utf8")) as RuntimeTelemetryStatus;
    return value.version === 1 ? value : null;
  } catch {
    return null;
  }
}

export function runtimeTelemetryPath(configPath: string): string {
  return runtimeArtifactPath(configPath, "status");
}

export function clearRuntimeTelemetry(configPath: string): void {
  try {
    unlinkSync(runtimeTelemetryPath(configPath));
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code !== "ENOENT") throw error;
  }
}

function credentialFreeRemoteLabel(remoteLabel: string): string {
  try {
    const url = new URL(remoteLabel);
    url.username = "";
    url.password = "";
    url.search = "";
    url.hash = "";
    return url.toString();
  } catch {
    return remoteLabel
      .replace(/^[^/@]+@([^:]+):/, "$1:")
      .replace(/[?#].*$/, "");
  }
}

function credentialFreeGatewayLabel(gatewayUrl: string): string {
  try {
    const url = new URL(gatewayUrl);
    if (url.protocol !== "http:" && url.protocol !== "https:") return "invalid gateway";
    const path = url.pathname === "/" ? "" : url.pathname;
    return `${url.protocol}//${url.host}${path}`;
  } catch {
    return "invalid gateway";
  }
}
